analytical_engine/apps/pregel/louvain/louvain_app_base.h

```diff
@@ -0,0 +1,327 @@
+      if (ctx.halt) {
+        LOG(INFO) << "super step " << current_super_step << " decided to halt.";
+      }
+      LOG(INFO) << "[INFO]: superstep: " << current_super_step
+                << " pass: " << current_iteration / 2
+                << " totalChange: " << totalChange;
+    } else if (ctx.halt) {
+      double actual_quality =
+          ctx.compute_context_.template get_aggregated_value<double>(
+              ACTUAL_QUALITY_AGG);
+      // after one pass if already decided halt, that means stage 1 yield no
+      // changes, so we halt stage 2.
+      if (current_super_step <= 14 || actual_quality <= ctx.prev_quality) {
+        // stage 2 halt, the louvain computation complete.
+        LOG(INFO) << "stage 2 halt, ACTUAL QUALITY: " << actual_quality;
+        ctx.compute_context_.set_superstep(SYNC_RESULT_STEP);
+        sync_community(frag, ctx, messages);
+        messages.ForceContinue();
+        return;
+      } else if (ctx.compute_context_.superstep() > 0) {
+        // stage 1 halt
+        LOG(INFO) << "super step: " << current_super_step
+                  << " decided to halt, ACTUAL QUALITY: " << actual_quality
+                  << " previous QUALITY: " << ctx.prev_quality;
+        // start stage 2 to compress community.
+        ctx.compute_context_.set_superstep(PASS_RESTART_STEP);
+        ctx.prev_quality = actual_quality;
+        ctx.change_history.clear();
+        ctx.halt = false;
+      }
+    }
+
+    // At the start of each pass, every alive node send to their
+    // communities node info, we active the node.
+    if (ctx.compute_context_.superstep() == PASS_RESTART_STEP) {
+      ForEach(inner_vertices, [&ctx](int tid, vertex_t v) {
+        if (ctx.GetVertexState(v).is_alived_community) {
+          ctx.compute_context_.activate(v);
+        }
+      });
+    }
+
+    ForEach(inner_vertices, [&frag, &ctx, this](int tid, vertex_t v) {
+      if (ctx.compute_context_.active(v)) {
+        LouvainVertex<fragment_t, vd_t, md_t> pregel_vertex;
+        pregel_vertex.set_context(&ctx);
+        pregel_vertex.set_fragment(&frag);
+        pregel_vertex.set_compute_context(&ctx.compute_context_);
+        pregel_vertex.set_vertex(v);
+        pregel_vertex.set_tid(tid);
+        auto& cur_msgs = (ctx.compute_context_.messages_in())[v];
+        this->program_.Compute(
+            grape::IteratorPair<md_t*>(
+                &cur_msgs[0],
+                &cur_msgs[0] + static_cast<ptrdiff_t>(cur_msgs.size())),
+            pregel_vertex, ctx.compute_context_);
+      } else if (ctx.compute_context_.superstep() == COMPRESS_GRAPH_STEP) {
+        ctx.GetVertexState(v).is_alived_community = false;
+      }
+    });
+
+    {
+      // Sync Aggregator
+      ctx.compute_context_.aggregate(CHANGE_AGG, ctx.GetLocalChangeSum());
+      ctx.compute_context_.aggregate(TOTAL_EDGE_WEIGHT_AGG,
+                                     ctx.GetLocalEdgeWeightSum());
+      ctx.compute_context_.aggregate(ACTUAL_QUALITY_AGG,
```

Comment:
Do you assume `Q` stands for `quality`? Seems not very approriate.

Reply:
in the louvain alogrithm origin code , the Q did represent quality i think